### PR TITLE
Book genres

### PR DESCRIPTION
**Pull request by Tymoteuszauke**
Added genres table and relation to book discounts.
Fixed bugs (Matras scrapper)
Fixed tests
Added flyway dependency into bot persistence module

### CODE REVIEW

File: matrasscraper/src/main/java/com/blackbook/matrasscraper/scraper/Scraper.java [hunk 1]
```diff
@@ -65,7 +65,7 @@ public void start(Consumer<List<BookDiscountData>> consumer) {
         Document mainPageDoc = htmlDocumentProvider.provide(collectorData.getBaseUrl());
         lastPageNo = Math.min(lastPageNo, extractLastPageNo(mainPageDoc));
         List<BookDiscountData> bookDiscountData = new LinkedList<>();
-        for (int i = 0; i < lastPageNo; i++) {
+        for (int i = 1; i < lastPageNo; i++) {
```
Comment (agawda) [inline matrasscraper/src/main/java/com/blackbook/matrasscraper/scraper/Scraper.java:68]: Why 1 and not 0 like before?

File: matrasscraper/src/test/resources/matras.html [hunk 2]
```diff
@@ -1,19 +1,29 @@
...
-            <span class="s-item-desc">
+                        <span class="s-item-desc">
                 <span class="s-counter no-counter">
                     <span class="count">1</span>
                     <span class="right-side">
-                        <a title="" href="http://www.matras.pl/smakowita-ella-cudowne-produkty-i-przepyszne-jedzenie-ktore-pokochacie-ty-i-twoje-cialo,p,248228"><span class="item-title">Smakowita Ella. Cudowne produkty i&nbsp;przepyszne jedzenie, któr...</span></a>
-                                                                                                <a title="" href="http://www.matras.pl/szukaj/?autor=Ella+Woodward"><span class="item-author">Ella Woodward</span></a>
+                        <a title=""
+                           href="http://www.matras.pl/smakowita-ella-cudowne-produkty-i-przepyszne-jedzenie-ktore-pokochacie-ty-i-twoje-cialo,p,248228"><span
+                                class="item-title">Smakowita Ella. Cudowne produkty i&nbsp;przepyszne jedzenie, któr...</span></a>
```
Comment (agawda) [inline matrasscraper/src/test/resources/matras.html:23]: 🥘

File: matrasscraper/src/main/java/com/blackbook/matrasscraper/scraper/Scraper.java (hunk 1 above)
Comment (Dzionek95) [inline matrasscraper/src/main/java/com/blackbook/matrasscraper/scraper/Scraper.java:68]: IntStream.range(0,max) would be better

File: persistencebot/pom.xml [hunk 3]
```diff
@@ -13,6 +13,11 @@
             <artifactId>utils</artifactId>
             <version>1.0-SNAPSHOT</version>
         </dependency>
+        <dependency>
+            <groupId>org.flywaydb</groupId>
+            <artifactId>flyway-core</artifactId>
+            <version>4.2.0</version>
```
Comment (agawda) [inline persistencebot/pom.xml:19]: Magic number

Comment (Dzionek95) [inline persistencebot/pom.xml:19]: magic number ;(

File: persistencebot/pom.xml [hunk 4]
```diff
@@ -13,6 +13,11 @@
             <artifactId>utils</artifactId>
             <version>1.0-SNAPSHOT</version>
```
Comment (Dzionek95) [inline persistencebot/pom.xml:14]: same here

File: persistencebot/src/main/java/com/blackbook/persistencebot/service/GenreService.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Autowired
+    public void setBooksRepository(BooksRepository booksRepository) {
+        this.booksRepository = booksRepository;
+    }
+
+    @Async
+    public void addGenreToDatabase(BookData bookData, Book book) {
+        Genre genre = genreRepository.findByName(bookData.getGenre());
+        if (genre == null) {
```
Comment (Dzionek95) [inline persistencebot/src/main/java/com/blackbook/persistencebot/service/GenreService.java:35]: Objects.isNull() would be better

File: persistencebot/src/test/java/com/blackbook/persistencebot/model/BookTest.java [hunk 6]
```diff
@@ -62,12 +62,12 @@ public void testToString(String title, String subtitle, String genre) {
...
         book1.setTitle(title);
         book1.setSubtitle(subtitle);
-        book1.setGenre(genre);
+//        book1.setGenre(genre);
 
         Book book2 = new Book();
         book2.setTitle(title);
         book2.setSubtitle(subtitle);
-        book2.setGenre(genre);
+//        book2.setGenre(genre);
```
Comment (agawda) [inline persistencebot/src/test/java/com/blackbook/persistencebot/model/BookTest.java:70]: 8 commented out lines in this class.

File: persistencebot/src/test/java/com/blackbook/persistencebot/controller/BookDiscountsControllerTest.java [hunk 7]
```diff
@@ -169,7 +175,7 @@ private void initDataForSavingTest() {
         Book book = new Book();
         book.setTitle(bookData.getTitle());
         book.setSubtitle(bookData.getSubtitle());
-        book.setGenre(bookData.getGenre());
+//        book.setGenre(bookData.getGenre());
```
Comment (Dzionek95) [inline persistencebot/src/test/java/com/blackbook/persistencebot/controller/BookDiscountsControllerTest.java:178]: comment in test :( :-1:

**Review by agawda — APPROVED** [review]

File: persistencebot/src/test/java/com/blackbook/persistencebot/model/BookTest.java [hunk 8]
```diff
@@ -28,12 +28,12 @@ public void testEquals(String title, String subtitle, String genre) {
...
```
Comment (Dzionek95) [inline persistencebot/src/test/java/com/blackbook/persistencebot/model/BookTest.java:37]: again and again

File: persistencebot/src/test/java/com/blackbook/persistencebot/model/BookTest.java [hunk 9]
```diff
@@ -62,12 +62,12 @@ public void testToString(String title, String subtitle, String genre) {
...
```
Comment (Dzionek95) [inline persistencebot/src/test/java/com/blackbook/persistencebot/model/BookTest.java:71]: :O

File: persistencebot/src/test/java/com/blackbook/persistencebot/util/ViewMapperUtilTest.java [hunk 10]
```diff
@@ -67,7 +67,7 @@ private BookDiscount getDiscountWithoutBookstore() {
         book.setTitle("Pan Tymek");
         book.setSubtitle("-");
         book.setAuthors("Tymke Wergiliusz");
-        book.setGenre("Biografia");
+//        book.setGenre("Biografia");
         book.setCoverUrl("www.covers.com/tymek");
```
Comment (Dzionek95) [inline persistencebot/src/test/java/com/blackbook/persistencebot/util/ViewMapperUtilTest.java:71]: Again and again

**Review by Dzionek95 — APPROVED** [review]

**Pull request merged**